### ISSUE TITLE
Fix the macro HMM data + training pipeline: degenerate states, look-ahead bias, and no validation gate

### ISSUE THREAD

**Issue by nevilcp** [nevilcp/argus#10]
Follow-up to #8, which activated the HMM and shipped the artifact this issue
replaces. Findings below come from re-fitting the model on a reconstruction of
its own training set (13,010 rows, matching the artifact's recorded metadata).

## Context

The shipped artifact `argus/models/macro_hmm.joblib` is degenerate: two of its
three hidden states are the same distribution (`|Δμ|max = 0.00097`, covariance
Frobenius diff 0.0017), and the transition matrix alternates between them every
day (dwell time 1.00). The consequence is that the reported regime is a function
of **window-length parity**, not of the macro data — `tail(89)` and `tail(90)`
on the same real window return TRANSITIONAL and EXPANSION respectively.
`MACRO.inference_window_days` is tagged ARBITRARY, so changing it by one flips
every EXPANSION in the system.

Measured behaviour of the current model:

| Check | Result |
|---|---|
| Goldilocks boom (VIX 11, FF 1.0, U 3.8) | **CONTRACTION, conf 1.000** |
| Deep 2023 curve inversion | **EXPANSION** |
| Stagflation (CPI 9%, inverted, VIX 35) | **EXPANSION** |
| 1992–2026 monthly trace | Oct-2008 → May-2017 labelled CONTRACTION (104 months, spanning the longest bull market in US history) |
| First 10 months of the 2007-09 recession | EXPANSION |
| Single-observation fallback path | Constant TRANSITIONAL — `startprob_[CONTRACTION] = 0.0` exactly, so CONTRACTION is unreachable |
| Cultural memory (207 stored rows) | 207/207 tagged EXPANSION — the field carries zero information |

Three root causes, all confirmed by re-fitting on the reconstructed training set:

1. **Daily forward-fill of monthly data.** 3 of 5 features are monthly
   (439/438/438 obs) but training resamples to calendar-daily → 13,010 rows.
   `fed_funds` carries new information on 2.6% of rows, `unemployment` 2.4%;
   29.9% of rows are exact duplicates; lag-1 autocorrelation is 0.998–0.9999.
   That flatly violates the HMM's conditional-independence assumption, and EM
   absorbs the persistence into a 2-cycle. Re-fitting the *same* model on the
   *same* data proves the causality:

   | Sampling | n | closest-pair \|Δμ\|max | dwell (obs) |
   |---|---|---|---|
   | daily ffill (shipped) | 13,010 | 0.0010 **degenerate** | 1.0, 1.0, 1436 |
   | business-day | 9,294 | 0.0014 **degenerate** | 1.0, 1.0, 1030 |
   | weekly | 1,859 | 0.0039 **degenerate** | 1.0, 1.0, 205 |
   | **monthly** | 428 | **1.63 ok** | 48.7, 8.7, 60.9 |

2. **Look-ahead bias.** FRED indexes by *reference* date, not publication date.
   Against publication-lagged series: `cpi_yoy` differs on 99.5% of rows,
   `fed_funds` 79.7%, `unemployment` 74.0% (max error 11.3pp — on 2020-04-15
   training sees UNRATE 14.8 when a real-time observer knew 4.4).

3. **Non-stationary level features + single fixed-seed EM.** Levels make the HMM
   segment *rate eras* rather than regimes (hence "CONTRACTION" = the ZIRP
   period). And 3 of 5 seeds collapse on the shipped configuration; seed 42 is
   one of them.

Nothing in the pipeline would have caught this: `train_macro_hmm.py:53` checks
only `is_fitted`, which a degenerate model passes, and `tests/test_macro.py`
fits on synthetic well-separated blobs while hand-patching `startprob_` — so it
validates the code path, not the model.

**Intended outcome:** a macro regime signal that is a function of the macro data,
trained only on information available in real time, with a validation gate that
fails the training run rather than shipping a broken artifact.

### Validated target design

Prototyped end-to-end before writing this plan. Monthly + point-in-time +
stationary features, `k=3`, best of 20 restarts with degeneracy rejection:

```
   months  recession_%  dwell_mo  d_fed_funds_6m  d_unemp_12m  cpi_yoy  t10y2y  vix_pctile
0     114          0.0      28.9            0.04        -0.62     1.80    1.73       23.90
1     122         22.1      24.3           -0.62         0.85     3.29    1.52       61.91   <- CONTRACTION
2     191          1.6      63.0            0.23        -0.25     2.74    0.25       51.22
```

Min pairwise Mahalanobis separation **2.71** (vs 0.001 shipped). NBER base rate
7.0% → contraction state **3.2× enriched, 90% recall** of all recession months.
Contraction spans: 1991-02→1993-01, 2001-05→2003-05, **2007-11→2010-06** (one
month *before* the NBER start), 2011-07→2011-12 (Euro crisis), 2020-05→2023-03.

---

## Stage 1 — Shared point-in-time feature construction

New module `argus/data/macro_features.py`. This is the single most important
change: training and inference currently build features through two independent
code paths (`fit_on_history` uses `yf.download` + `fetch_fred_series` directly;
`_assemble_feature_window` uses the `MarketDataProvider` seam), with nothing
asserting they agree.

- Add `PUBLICATION_LAG_DAYS` to `argus/params.py` as a new provenance-tagged
  `MacroParams` field group (`Provenance.LITERATURE` — these are published BLS/
  Fed release calendars): `FEDFUNDS: 32`, `UNRATE: 35`, `CPIAUCSL: 45`.
- `build_macro_feature_frame(provider: MarketDataProvider, start: str) -> pd.DataFrame`:
  1. Pull `FEDFUNDS`, `UNRATE`, `CPIAUCSL`, `T10Y2Y` via `provider.fred_series`
     and `^VIX` via `provider.ohlcv_daily` — **no direct fetcher calls**, so the
     `FixtureMarketDataProvider` seam works for both paths.
  2. Shift each monthly series' index by its publication lag.
  3. Derive the stationary features, then `resample("ME").last()`:
     `d_fed_funds_6m` = `.diff(6)`, `d_unemp_12m` = `.diff(12)`,
     `cpi_yoy` = `.pct_change(12) * 100`, `t10y2y` = level (mean-reverting),
     `vix_pctile` = 5-year rolling percentile rank (`rolling(1260, min_periods=252).rank(pct=True) * 100`).
  4. `.ffill(limit=2).dropna()` — a bounded fill tolerates a late release
     without silently carrying a stale print for a year.
- Replace `FEATURE_COLUMNS` in `argus/agents/macro.py:50` with the five names
  above. Keep the existing artifact-metadata mismatch check — it will now
  correctly reject the old artifact at load.
- `_FEATURE_DEFAULTS` (`macro.py:52`) must be re-keyed to the new features.

Delete `MacroStatisticalAgent.fit_on_history`'s bespoke fetching
(`macro.py:321-367`) and `_assemble_feature_window`'s (`macro.py:369-412`); both
become thin calls to `build_macro_feature_frame`. This removes the
`resample("D").asfreq().ffill()` construction entirely (both copies).

`fetch_ohlcv_daily(period="2y")` is not enough history for a 5-year VIX
percentile — the feature builder needs `period="10y"` (or `max` for training).

## Stage 2 — Inference mechanics

In `argus/agents/macro.py`:

- **Stationary `startprob_`.** In `RegimeClassifier.load`, after restoring the
  HMM, overwrite `startprob_` with the transition matrix's stationary
  distribution (left eigenvector for λ=1, via `scipy.linalg.eig` — scipy is
  already a dependency). Fitting one long sequence gives hmmlearn nothing to
  average over, so `startprob_` collapses to a one-hot artifact of wherever the
  sequence began. `tests/test_macro.py` already documents this and patches it;
  production must do the same. This is what makes CONTRACTION reachable from the
  single-observation path.
- **`MACRO.inference_window_days` → `inference_window_months`**, default 36
  (3 years of monthly observations; keep `Provenance.ARBITRARY` and note it is
  not backtested).
- **`analyze()` populates `MacroContext` from the window's final row**, not from
  a separately-fetched `macro_bundle()` (`macro.py:433, 447-449`). Two
  independent fetches can disagree. Keep the daily VIX path
  (`macro.py:464-480`) exactly as-is — `vix_level`, `vix_percentile`,
  `vix_regime` and `agent_multipliers` stay daily-responsive, which is where
  daily responsiveness belongs. The regime itself only moves monthly.
- **Surface model health.** Add `model_healthy: bool` to `MacroContext`
  (`argus/schemas/signals.py:113`). Set it `False` when the classifier fell back
  to rules, so downstream consumers can distinguish a real posterior from the
  fixed 0.5/0.6 heuristic confidence — today they are indistinguishable.
- **Loosen the version check** (`macro.py:264-273`) from exact string equality
  to major.minor. A `pip install -U scikit-learn` currently drops production to
  the rule-based path silently. Also floor `scikit-learn` and `hmmlearn` in
  `pyproject.toml:30-31` to the trained majors.
- `predict()` should reject a window containing NaN/inf rather than letting it
  propagate to a silent `argmax` on NaN.

## Stage 3 — Fit protocol

Rewrite `RegimeClassifier.fit` (`macro.py:86-104`):

- Fit `StandardScaler` on the monthly frame (no duplicated rows — the current
  scaler's mean/std are weighted by calendar days, counting Fridays 3× via
  weekend fill).
- **`n_init` restarts** (default 20) over `random_state=0..n_init-1`, keeping the
  highest `score(X)`.
- **Reject degenerate solutions**: minimum pairwise Mahalanobis distance between
  state means (pooled covariance, `scipy.spatial.distance.mahalanobis`) must
  exceed a threshold — 1.0 is a safe floor; the validated fit scores 2.71 and
  the shipped one ~0.001. Also require every state to hold ≥5% of observations.
- **`fit` must raise** when no restart passes, instead of leaving `is_fitted`
  False for the caller to notice.
- Keep `n_components=3` fixed. BIC decreases monotonically to k=5 on this sample
  (3968 → 3619 → 3448), but k>3 forces an arbitrary many-to-3 mapping onto the
  `Regime` enum for no validated gain; k=3 already achieves 3.2× NBER
  enrichment. Log BIC for k∈{2..5} as a diagnostic so the choice stays visible.
- hmmlearn prints "Model is not converging" on sub-1e-4 monotonicity wobble near
  the optimum; that is benign, do not treat it as a failure.

## Stage 4 — Labelling and the validation gate

Replace `_map_states` (`macro.py:106-148`). The current rule assigns CONTRACTION
to `means["vix"].idxmax()`, so exactly one state is *always* CONTRACTION whether
or not a contractionary state exists — the model cannot express "no contraction
here", which is how the ZIRP era got the label.

- Fetch `USREC` (NBER recession indicator, on FRED) over the training window and
  compute each state's recession frequency.
- CONTRACTION = the state with the highest recession frequency, **and** it must
  clear a minimum enrichment ratio vs the sample base rate (require ≥2.0×; the
  validated fit achieves 3.2×). If no state clears it, fail the run.
- EXPANSION = of the remaining states, the one with the lowest `d_unemp_12m`
  (falling unemployment) — economic content, not a VIX rank.
- Everything else TRANSITIONAL. Clear `state_to_regime` at the start of each fit
  (it is currently never reset between fits).
- Persist the validation metrics (per-state recession frequency, enrichment,
  recall, min separation, dwell times) into the artifact's `metadata` so a
  shipped model carries its own evidence.

Then add the gate to `scripts/train_macro_hmm.py`, which today only checks
`is_fitted` (line 53). Hard-fail with a non-zero exit on: min state separation
below threshold; any state under the occupancy floor; dwell time outside 3–120
months; CONTRACTION enrichment under 2.0×; or a **discrimination test** — run a
set of canned scenarios (GFC crash, COVID, 2023 inversion, Goldilocks,
stagflation) and fail if they do not produce at least two distinct labels. That
last check alone would have caught the shipped artifact.

Also change the script's `--start-date` default from `2010-01-01` to
`1990-01-01`, which is what the shipped artifact was actually trained with — the
documented default and the real invocation currently disagree.

## Stage 5 — Tests

`tests/test_macro.py` needs tests that would fail against the current artifact:

- **Window-length invariance**: the same constant window at lengths n and n+1
  must return the same regime. This is the direct regression test for the
  parity bug.
- **Discrimination**: a crash-like window and a boom-like window must not
  produce the same label.
- **CONTRACTION reachable from a single observation** (guards the `startprob_`
  fix).
- **Degeneracy rejection**: construct data with a duplicated cluster, assert
  `fit` raises rather than returning a degenerate model.
- **Point-in-time**: assert no feature row contains information published after
  that row's timestamp — build a fixture with a known release date and assert it
  is absent from earlier rows.
- **Train/serve parity**: `build_macro_feature_frame` against the same
  `FixtureMarketDataProvider` must produce identical frames for the training and
  inference call sites.
- Update `_fit_synthetic_classifier` to stop hand-patching `startprob_`; the
  production `load` path now handles it, and the patch is currently masking the
  bug from the suite.

## Stage 6 — Remediation and retrain

1. Retrain: `.venv/bin/python -m scripts.train_macro_hmm --start-date 1990-01-01`.
   Commit the new artifact only if the gate passes.
2. One-off script (`scripts/` , not wired into any schedule) rewriting the
   cultural-memory regime tag to `unknown` on all 207 rows in
   `chroma_db/chroma.sqlite3` — both the `embedding_metadata` row and the
   `Macro regime:` line in the document text written at `cultural.py:119,317`.
   All 207 currently read EXPANSION, so no information is lost. Back up the
   sqlite file first.
3. `cultural.py:164,196` builds retrieval queries and a `$and` filter on
   `regime`; make it skip the regime clause when the value is `unknown`, or the
   filter will match nothing.

`argus/backtesting/replay.py:154` passes `macro_context=None`, so backtesting is
unaffected and needs no changes.

---

## Verification

1. `.venv/bin/pytest tests/test_macro.py -v` — new invariance, discrimination,
   degeneracy and point-in-time tests pass.
2. `.venv/bin/pytest` — full suite; `test_golden_dag.py` and `test_integration.py`
   are the likely breakages from the `MacroContext` field addition.
3. `.venv/bin/python -m scripts.train_macro_hmm --start-date 1990-01-01` — gate
   passes; printed per-state table should reproduce ~3.2× NBER enrichment, ~90%
   recall, min separation ≳2.5.
4. **Regression harness for the original bug** — assert the regime is stable
   across window lengths and that the scenario sweep produces distinct labels:
   ```
   for n in range(34, 39):  # months
       assert clf.predict(window.tail(n))[0] == expected
   ```
5. Re-run the 1992–2026 monthly trace and confirm CONTRACTION spans overlap NBER
   recessions instead of the ZIRP era.
6. `.venv/bin/ruff check .` and `.venv/bin/mypy argus` — line-length 100, and
   every new public `def`/`class` needs the Google-style docstring plus the
   module docstring shape from `CLAUDE.md`.
7. Confirm `select regime, count(*)` over the ChromaDB metadata returns
   `unknown: 207` after the remediation script.

## Suggested PR split

- **PR 1 (Stages 1–2)** — feature construction + inference mechanics. Root
  cause, mechanical, high confidence.
- **PR 2 (Stages 3–5)** — fit protocol, labelling, gate, tests. Contains the
  modelling judgement calls and deserves its own review.
- **PR 3 (Stage 6)** — retrained artifact + memory remediation.

**Comment by nevilcp** [comment]
Resolved via #14 (Stage 6), building on the Stage 1-5 work already merged in #13. Retrained the macro HMM on the point-in-time monthly pipeline — validation gate passes with 427 observations, min state separation 2.75, CONTRACTION enrichment 3.8x, 90% recall; CONTRACTION spans now track real NBER recessions instead of the ZIRP era. Also rewrote the 207 stale EXPANSION-tagged cultural-memory rows to unknown and made retrieval skip the regime clause when unknown.
